crates/revm/src/db/states/state.rs

```diff
@@ -547,4 +547,126 @@ mod tests {
             "The latest state of the existing account is incorrect"
         );
     }
+
+    /// Checks that the behavior of selfdestruct within the block is correct.
+    #[test]
+    fn selfdestruct_state_and_reverts() {
+        let mut state = StateBuilder::default().build();
+
+        // Existing account.
+        let existing_account_address = B160::from_slice(&[0x1; 20]);
+        let existing_account_info = AccountInfo {
+            nonce: 1,
+            ..Default::default()
+        };
+
+        let (slot1, slot2) = (U256::from(1), U256::from(2));
+
+        // Existing account is destroyed.
+        state.apply_transition(Vec::from([(
+            existing_account_address,
+            TransitionAccount {
+                status: AccountStatus::Destroyed,
+                info: None,
+                previous_status: AccountStatus::Loaded,
+                previous_info: Some(existing_account_info.clone()),
+                storage: HashMap::default(),
+                storage_was_destroyed: true,
+            },
+        )]));
+
+        // Existing account is re-created and slot 0x01 is changed.
+        state.apply_transition(Vec::from([(
+            existing_account_address,
+            TransitionAccount {
+                status: AccountStatus::DestroyedChanged,
+                info: Some(existing_account_info.clone()),
+                previous_status: AccountStatus::Destroyed,
+                previous_info: None,
+                storage: HashMap::from([(
+                    slot1,
+                    StorageSlot {
+                        previous_or_original_value: U256::ZERO,
+                        present_value: U256::from(1),
+                    },
+                )]),
+                storage_was_destroyed: false,
+            },
+        )]));
+
+        // Slot 0x01 is changed, but existing account is destroyed again.
+        state.apply_transition(Vec::from([(
+            existing_account_address,
+            TransitionAccount {
+                status: AccountStatus::DestroyedAgain,
+                info: None,
+                previous_status: AccountStatus::DestroyedChanged,
+                previous_info: Some(existing_account_info.clone()),
+                // storage change should be ignored
+                storage: HashMap::from([(
+                    slot1,
+                    StorageSlot {
+                        previous_or_original_value: U256::from(1),
+                        present_value: U256::from(2),
+                    },
+                )]),
```

Comment:
Destroyed account (`DestroyedAgain` here) will not have storage

Selfdestruct is called here: https://github.com/bluealloy/revm/blob/main/crates/revm/src/db/states/cache.rs#L102
And transition created here:
https://github.com/bluealloy/revm/blob/main/crates/revm/src/db/states/cache_account.rs#L241

We should maybe be more explicit about this set storage to empty hashmap here: https://github.com/bluealloy/revm/blob/main/crates/revm/src/db/states/transition_account.rs#L68